File: 0980-unique-paths-iii/0980-unique-paths-iii.py
class Solution:
    def uniquePathsIII(self, grid: List[List[int]]) -> int:    
        self.ans = 0
        m, n, empty = len(grid), len(grid[0]), 1
        dirs = [[1, 0], [0, 1], [-1, 0], [0, -1]]
        
        for i in range(m):
            for j in range(n):
                if not grid[i][j]:
                    empty += 1
                elif grid[i][j] == 1:
                    x, y = i, j
                    
        def dfs(x, y, empty):
            if x < 0 or x >= m or y < 0 or y >= n or grid[x][y] < 0:
                return
            
            if (grid[x][y] == 2):
                self.ans += empty == 0
                return
              
            grid[x][y] = -2
            empty -= 1
            
            for dx, dy in dirs:
                dfs(dx + x, dy + y, empty)
            
            grid[x][y] = 0
            empty += 1
            
                    
        dfs(x, y, empty)
        
        return self.ans
        
        

File: 0980-unique-paths-iii/0980-unique-paths-iii.js
/**
 * @param {number[][]} grid
 * @return {number}
 */
var uniquePathsIII = function(grid) {
  const [m, n] = [grid.length, grid[0].length];
  const dirs = [[1, 0], [0, 1], [-1, 0], [0, -1]];
  let [empty, ans, x, y] = [1, 0, -1, -1];
  
  const dfs = (r, c) => {
    if (r < 0 || r >= m || c < 0 || c >= n || grid[r][c] < 0) return;
    
    if (grid[r][c] === 2) {
      if (!empty) ans++;
      return;
    }
    
    grid[r][c] = -2;
    empty--;
    
    dirs.forEach(([dx, dy]) => dfs(r + dx, c + dy));
    
    grid[r][c] = 0;
    empty++;
  }
  
  for (let i = 0; i < m; i++) {
    for (let j = 0; j < n; j++) {
      const curr = grid[i][j];
      if (grid[i][j] === 1) [x, y] = [i, j];
      if (!grid[i][j]) empty++;
    }
  }
  
  dfs(x, y);
  
  return ans;
};